### ISSUE TITLE
Polygon and Marker issues with Leaflet.Draw version 0.4.9

### ISSUE THREAD

**Issue by Searlon** [arthur-e/Wicket#103]
I've been upgrading the leaflet and leaflet plug in libraries, along with Wkt.  I was originally using the following versions and everything worked fine:
Leaflet 0.7.5
Leaflet.Draw 0.2.4
Wkt 1.3

However, now I'm upgrading to the latest versions:
Leaflet 1.0.2
Leaflet.Draw 0.4.9
Wkt 1.3.2

What I've found is that Polygons will be parsed as Wkt without any errors, however their latitude and longitude values are undefined.  
Markers will not be parsed by the Wkt library and will throw an error.  I've done some digging on this one and it turns out Leaflet.Draw has changed the constructor for the Marker from L.Marker to L.Marker.Touch.

See this jsfiddle for simple scenarions, just draw a polygon or place a marker: https://jsfiddle.net/s057n0ze/

**Comment by arthur-e** [comment]
To clarify, this problem only arises when Leaflet.Draw is used?

**Comment by Searlon** [comment]
I've managed to replicate it in this JSFiddle without using Leaflet.Draw: https://jsfiddle.net/ve4Lkdsh/

**Comment by arthur-e** [comment]
Thanks for this detailed report. Hopefully, someone can identify the problem and issue a pull request soon.

**Comment by Searlon** [comment]
I suppose I should clarify a little more as my original post mentions issues about Polygons and Markers.

The issue for Polygons happens without using Leaflet.Draw, as demonstrated in the above JSFiddle.

However the issue with Markers only happens when placing a Marker when using Leaflet.Draw.  This is because Leaflet.Draw has changed the constructor for the Marker from L.Marker to L.Marker.touch.

**Comment by jblamber** [comment]
+1 am also seeing this issue. Wicket is correctly converting circles, lines and squares to WKT objects, bu the polygon just shows up as POLYGON((undefined undefined,undefined undefined)) - irrespective of how many points make up the polygon. The same as OP - My polygons are created from leaflet draw. Will investigate further.

**Comment by jblamber** [comment]
This is the reverse issue of #95 - when the WKT parser tries to extract the latLngs from a polygon (which can now also be a multipolygon) its recieving a two dimensional array

Polyline.js line 59+
`// @method getLatLngs(): LatLng[]
	// Returns an array of the points in the path, or nested arrays of points in case of multi-polyline.
	getLatLngs: function () {
		return this._latlngs;
	},`

Which results in the verts object in the L.polygon class getting a bunch of undefined points. verts needs to become a nested array too. Thus the L.Polygon parser will need to be upgraded to support multipolygons too:

Wicket-leaflet-.js line 276+
`// L.Polygon ///////////////////////////////////////////////////////////////

    if (obj.constructor === L.Polygon || obj.constructor === L.polygon) {
        rings = [];
        verts = [];
        tmp = obj.getLatLngs();

        // First, we deal with the boundary points
        for (i = 0; i < obj._latlngs.length; i += 1) {
            verts.push({ // Add the first coordinate again for closure
                x: tmp[i].lng,
                y: tmp[i].lat
            });
        }

        verts.push({ // Add the first coordinate again for closure
            x: tmp[0].lng,
            y: tmp[0].lat
        });
`

This is just part of a hobby project of mine. Will try my hand at a pull request over the next while - but someone else might beat me :)

**Comment by jblamber** [comment]
More investigation - the way circles/ "rectangles" are handled from leaflet is now not 100% correct. Circles now use a L.Circle constructor (apparantly used a L.Rectangle before?) 

```
 if (obj.constructor === L.Rectangle || obj.constructor === L.rectangle) {
        console.log('Deconstruction of L.Circle objects is not yet supported');

    } else {
        console.log('The passed object does not have any recognizable properties.');
    }
```

Rectanges are now just L.Polygons (single dimension) but "polygons" are now multidimensional.

**Comment by jblamber** [comment]
If you need a fast fix for leaflet-draw (which doesnt seem to support polygon holes - you can use this inside wicket-leaflet.js starting @ line 276

Over the next few days ill try to source some samples with polygon holes to test that. But this works for the basic Polygons leaflet now uses while being backwards compatible.

```
 // L.Polygon ///////////////////////////////////////////////////////////////

    if (obj.constructor === L.Polygon || obj.constructor === L.polygon) {
        rings = [];
        tmp = obj.getLatLngs();

        if (tmp.constructor === Array) {
            for (ring = 0; ring < tmp.length; ring += 1) {
                verts = [];
                // First, we deal with the boundary points
                for (i = 0; i < tmp[ring].length; i += 1) {
                    verts.push({ // Add the coordinate
                        x: tmp[ring][i].lng,
                        y: tmp[ring][i].lat
                    });
                }
                verts.push({ // Add the first coordinate again for closure
                    x: tmp[ring][0].lng,
                    y: tmp[ring][0].lat
                });
                rings.push(verts);
            }

        }
        else {
            verts = [];
            // First, we deal with the boundary points
            for (i = 0; i < obj._latlngs.length; i += 1) {
                verts.push({ // Add the coordinate
                    x: tmp[i].lng,
                    y: tmp[i].lat
                });
            }

            verts.push({ // Add the first coordinate again for closure
                x: tmp[0].lng,
                y: tmp[0].lat
            });

            rings.push(verts);

            // Now, any holes
            if (obj._holes && obj._holes.length > 0) {
                // Reworked to support holes properly
                verts = coordsFromLatLngs(obj._holes);
                for (i = 0; i < verts.length; i++) {
                    verts[i].push(verts[i][0]); // Copy the beginning coords again for closure
                    rings.push(verts[i]);
                }
            }
            
        }
        return {
            type: 'polygon',
            components: rings
        };
    }
```

**Comment by arthur-e** [comment]
Great, thanks @jblamber! Can you put this in a pull request?

**Comment by arthur-e** [comment]
Haven't received an update in ~2.5 years, so I'm hoping that this was fixed in PR #121. Can re-open if the issue persists.
